Endpoints for the new UI

anttimaki opened the pull request.

MythicManiac commented on django/thunderstore/frontend/api/experimental/views/frontpage.py:
```diff
@@ -0,0 +1,104 @@
+    return result["downloads__sum"] or 0
+
+
+def get_mods_grand_total() -> int:
```
I would combine this & `get_downloads_grand_total` to a single query:

```python
result = Package.objects.annotate(count=Value(1)).aggregate(
    downloads_total=Sum("downloads"),
    count=Sum("count", output_field=models.BigIntegerField()),
)
```

You might even be able to skip the annotate and just do


```python
result = Package.objects.aggregate(
    downloads_total=Sum("downloads"),
    count=Value(1, output_field=models.BigIntegerField()),
)
```

MythicManiac commented on django/thunderstore/frontend/api/experimental/views/frontpage.py:
```diff
@@ -0,0 +1,104 @@
+        community_objs = Community.objects.listed().prefetch_related(
+            "package_listings__package__versions",
+            "sites",
+        )
```
The output field of BigIntegerField might not be necessary for now but it's future proofing.

MythicManiac commented on django/thunderstore/frontend/api/experimental/views/frontpage.py:
```diff
@@ -0,0 +1,104 @@
+            }
+            for c in community_objs
+        ]
+        communities.sort(key=lambda c: c["modCount"], reverse=True)
```
If you annotate the package count, you can use a queryset `.order_by("-package_count")` here as well.

MythicManiac reviewed (changes requested):
See comments

anttimaki commented:
Based on the earlier comments, [the second commit](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/1d7d15291d73b0e70eedf44719468e8475f614ea) has changed significantly:

* Use annotations to fetch **community-specific** package/download counts
* Use sum() in Python to calculate **grand total** package/download counts
* Due to the above changes, most of the helper functions in `frontpage.py` were no longer needed
* Due to removing the helper functions, tests needed to be refactored completely
* Since `get_site_image` was the only helper left, it was refactored to `Community`'s property, see [the new first commit](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/dbc93f8b6c5cf5c318b9d796159b2cfb272978d2)
* Using `serializer_class` property in the `FrontPageApiView` made little sense since `APIView`s don't actually seem to use it, so it was removed and direct references are used instead
* Refactored `FrontPageApiView` to use `get_queryset` and `serialize_results` methods, consistent with the two other views in this PR

Additionally three new commits were added: two for new endpoints, and one for a decorator which I used while working the ORM queries to ensure no unnecessary database queries were made.

MythicManiac commented on django/thunderstore/frontend/api/experimental/tests/test_package_details.py:
```diff
@@ -0,0 +1,192 @@
+
+    data = __query_api(api_client, dependant)
+
+    assert len(data["dependencies"]) == 1
```
I feel like this test should assert the two package listings are in different communities somehow, otherwise we don't really know if it's falling back to the non-preferred community or if they just happen to be in the same community

anttimaki replied:
The way the dependant/dependency are initiated with parameterless factories guarantees they're in different communities, but that's not visible from reading the test case, and someone can always change how the factory works, so I'll add a check just in case.

MythicManiac commented on django/thunderstore/frontend/api/experimental/tests/test_package_details.py:
```diff
@@ -0,0 +1,192 @@
+
+    data = __query_api(api_client, dependant)
+
+    assert len(data["dependencies"]) == 1
```
Likewise here I'd expect assertions to be made about which dependencies are or aren't in the same community.

anttimaki replied:
Likewise.

MythicManiac commented on django/thunderstore/frontend/api/experimental/tests/test_package_details.py:
```diff
@@ -0,0 +1,192 @@
+
+
+@pytest.mark.django_db
+def test_rating_is_annotated_correctly_when_ratings(api_client: APIClient) -> None:
```
I'd almost suggest combining this and `test_rating_is_annotated_correctly_when_no_ratings` to a single parametrized test case, which will call `PackageRatingFactory` the amount of times defined in the parametrized decorator, and assert the rating score matches that count.

Standalone tests are fine too, but if you want to familiarize yourself with parametrized test cases, this would be a good example case.

anttimaki replied:
I'm familiar with parametrized tests, but when writing the tests I considered these two separate, "named" test cases. But I can see the argument for them not being separate too, so changing it in this instance.

Is this something that we want to have a "house rule" or project-specific way of doing things?

MythicManiac replied:
I don't think it's necessary here (or should be forced), but this is a borderline case where either approach would work. Named tests can be better in some scenarios to make it easier to see what fails, as they tend to be more verbose in the names for the specific failure reason.

Primary reason I commented on it is because I hadn't seen much use of parameterized tests and I wasn't sure if that was a conscious decision or just because it wasn't a habit.

In other words, if it was a conscious decision to _not_ use a parametrized test here, but it was considered, that's completely fine, just wanted to make sure it indeed was considered.

anttimaki replied:
Well, in this case it was *not* considered 😄 In my head "stuff happens when x exists" and "other stuff happens when x doesn't exist" were so clearly different things that they should be tested in separate test cases. But as I said, I can see that "correct stuff happens with different counts of x" is a valid and logical approach as well, and as a bonus provides more than just the binary test cases. And I also implemented this already soooo...

MythicManiac commented on django/thunderstore/frontend/api/experimental/tests/test_package_details.py:
```diff
@@ -0,0 +1,192 @@
+    PackageVersionFactory,
+)
+
+
```
I'd expect tests for approval status visibility rules to be included

anttimaki replied:
Adding tests.

MythicManiac commented on django/thunderstore/frontend/api/experimental/tests/test_community_package_list.py:
```diff
@@ -0,0 +1,418 @@
+
+    data = __query_api(api_client, site.community.identifier, "page=2")
+
+    assert len(data["packages"]) == 1
```
Perhaps could use a check making sure the package on page 2 doesn't also get listed on page 1? I've seen some weird pagination errors with APIs where multiple pages return same data (not on this project), so it feels like a sane check to have.

anttimaki replied:
Augmenting the test case.

MythicManiac commented on django/thunderstore/frontend/api/experimental/tests/test_community_package_list.py:
```diff
@@ -0,0 +1,418 @@
+    PackageVersionFactory(
+        package=listing1.package, downloads=9001, version_number="1.0.1"
+    )
+    bust_cache_by_condition(CacheBustCondition.any_package_updated)
```
could add an assert between new package version being created and cache being busted which ensures the cache to actually be working and stale data _is_ being returned _until_ the cache is busted.

MythicManiac replied:
I feel like it should be in this test too, as otherwise it's hard to reason (via assertions) on why the cache bust is required here. Imagine if a change is made where the cache busting is no longer necessary, IMO that would warrant removing the cache bust on this test, but if it's not asserted for it won't happen.

anttimaki replied:
Fair enough.

MythicManiac commented on django/thunderstore/frontend/api/experimental/tests/test_community_package_list.py:
```diff
@@ -0,0 +1,418 @@
+    data = __query_api(
+        api_client,
+        site.community.identifier,
+        f"included_categories={cat1.slug}&included_categories={cat2.slug}",
```
It dawned on me that this is indeed how Django does list query args by default. We might want to use comma separated slugs with a single key instead, but I'm leaving that under the same "acceptable in experimental" reasoning as with the other query parameter related feedback

anttimaki replied:
Any specific reasons why comma-separation would be better?

MythicManiac replied:
It's just nicer to work with on the clientside quite often (anecdotal evidence), since this form of query parameter is more complex. I'd prefer a single key per field, and comma separated list seems the cleanest way to achieve that, though I'm open to other syntax suggestions.

Thinking about the practical options, it's probably easiest to use list field in conjunction with a `ModelChoiceField`, saving a lot of parsing on the view itself. You should be able to use the same pattern as the `PackageUploadMetadataSerializer` here:

```python
    categories = serializers.ListField(
        child=CommunityFilteredModelChoiceField(
            queryset=PackageCategory.objects.all(),
            to_field="slug",
        ),
        allow_empty=True,
    )
 ```

And then encode them to the URL as `categories=[slug1, slug2, slug3]`

MythicManiac replied:
Also now that I think about it, using a model choice field might be counter-intuitive for query usage, as we don't necessarily need the category objects, just the slugs. In any case I feel like that should illustrate the design options available quite well.

MythicManiac replied:
The query parameter parsing was changed, but the test still uses the multi-key format. Is this intentional?

anttimaki replied:
Well, there was no intent one way or another behind leaving this as it was. However, the changes affected how e.g. boolean fields are parsed from text to proper types, not how lists are submitted. The deserializer still eats a `QueryDict`, which expects lists to be passed in this way. Otherwise we would have to manually split the received string by the separator character.

MythicManiac replied:
We could delegate this to a separate issue since it's less about implementation detail and more about API design. Let's not block this PR with it at the very least.

MythicManiac replied:
Created TS-258 tracking it

MythicManiac commented on django/thunderstore/frontend/api/experimental/tests/test_community_package_list.py:
```diff
@@ -0,0 +1,418 @@
+
+
+@pytest.mark.django_db
+def test_only_community_listings_for_correct_community_are_prefetched(
```
How does this actually test it's a prefetch that happens, or does it not do that and the name is misleading?

Query count assertions can be done if found useful, there are some example tests with those.

anttimaki replied:
Renamed the test from `...are_prefetched` to `...are_included_in_queryset` to better communicate that this test case only tests the view's `get_queryset` method rather than HTTP response of the view.

MythicManiac reviewed (changes requested):
Nice work, seems like a solid implementation, especially the query optimizations seem well done.

At least one actual issue was found, but beyond that most comments are either very minor issues or just questions/discussion points.

anttimaki commented on django/thunderstore/cache/cache.py:
```diff
@@ -23,6 +23,11 @@ class CacheBustCondition(ChoiceEnum):
     dynamic_html_updated = "dynamic_html_updated"
 
 
+def bust_cache_by_condition(condition: CacheBustCondition) -> None:
```
No reason, just couldn't find the existing implementation. Willfix 👍🏻

anttimaki commented on django/thunderstore/community/factories.py:
```diff
@@ -27,3 +29,71 @@ class Meta:
+    @factory.lazy_attribute
+    def community(self):
+        """
+        To use an existing community, pass it via community_ parameter.
```
Well, no solutions pop in mind right now. I guess this is a more generic problem though, how to get the developers know what's already available (as seen with the cache busting celery task in the earlier comment). And requiring explicit `None` will indeed get cumbersome especially for the factories that support this for multiple relations.

MythicManiac replied:
Marking as resolved as this is on the meeting topics/todo list now

anttimaki commented on django/thunderstore/community/factories.py:
```diff
@@ -27,3 +29,71 @@ class Meta:
+        if self.package_:
+            return self.package_
+
+        ver = PackageVersionFactory()
```
I have a vague recollection that there were some cases where that didn't work. My best guess was that it related to passing relations, but as I tested them now, they seem to work as expected. So I'll change this.

anttimaki replied:
Well now that I read the next comment regarding the `setattr`, my recollection was probably about updating the related objects automatically created by factory calls.

anttimaki commented on django/thunderstore/community/factories.py:
```diff
@@ -27,3 +29,71 @@ class Meta:
+
+        if self.community_kwargs:
+            for attr, value in self.community_kwargs.items():
+                setattr(site.community, attr, value)
```
For one thing, we could compare the fields in `community_kwargs` against `Community._meta.fields`.

But I guess in the long run it would be better to change the other factories to support this (i.e. first call `CommunityFactory` with the `community_kwargs`, and then call `CommunitySiteFactory` and pass the newly created `Community` to it.

MythicManiac replied:
Agreed, but sounds like a change that should be done in a separate PR. Created an issue about it & closing this.

anttimaki commented on django/thunderstore/frontend/api/experimental/serializers/views.py:
```diff
@@ -0,0 +1,112 @@
+
+class CommunityPackageListSerializer(serializers.Serializer):
+    """
+    Data shown on Community's package list view on frontend.
```
Are they useful? Well, the codebase for backend is not bite-sized. It can be difficult to find what you're looking for, and it can be difficult to deduce whether what you have found is what you were looking for. Let's say a CLI developer stumbles upon this serializer. Maybe the docstring tells them that this frontend stuff probably isn't what they're looking for? 🤷🏻

That being said, maybe the docstrings in this file don't add that much more value compared to the class name and file path (which tells this is frontend stuff). But maybe it's in more easily digested format? Anyways, if we feel the docstrings don't currently add value, I'd rather consider supplementing them rather than removing them.

About the comment getting stale: since the serializer is used for one specific view in frontend, probably only way it getting stale is that the view is removed from the frontend, meaning that would make the code itself also stale, which renders the issue of having stale comments irrelevant in comparison. (Alternatively it could get stale if the same serializer is used in multiple views, but that's less likely with our "fetch all you need and only things you need for this specific view" approach.)

MythicManiac replied:
I suppose what we should do is consider what the goals of docstrings we write are and when to write them. As you mentioned, in this case what the docstring documents is _mostly_ self-evident from the class & module path/names. My primary concern is that docstrings aren't really tied to any static analysis, which tends to make them outdated, although I do see your point as well. Really this comes back to documentation practices & maintenance process.

In short, we probably don't need to do anything about the docstrings here and can keep them as is, but we might want to stop to consider how to handle them in the future.

anttimaki commented on django/thunderstore/utils/decorators.py:
```diff
@@ -10,3 +11,34 @@ def wrapper(*args, **kwargs):
     return wrapper
+
+
+def print_query_stats(fn: Callable[..., None]) -> Callable[..., None]:
```
Moved to `thunderstore.debug.decorators`. What's the correct way to ignore files/paths in this repo? Creating a `codecov.yml` file on the root?

MythicManiac replied:
There's a `.coveragerc` file in the code project root that has some exclude filters in place, adding it there should suffice

anttimaki commented on django/thunderstore/frontend/api/experimental/views/community_package_list.py:
```diff
@@ -0,0 +1,253 @@
+    )
+    def get(self, request: HttpRequest, **kwargs) -> HttpResponse:
+        communities = Community.objects.prefetch_related("package_categories")
+        community = get_object_or_404(communities, identifier=kwargs.get("community"))
```
I'm not sure what you're actually suggesting here.

anttimaki replied:
TBH I see very little benefit from this regarding typing, since Django's URL dispatcher guarantees it to be a non-empty string, but I guess it can improve readability. I'll change the new experimental views to use this approach, although they will then work differently from the rest of the views (at least the ones I checked).

MythicManiac replied:
This would be primarily for static checking for downstream use, as the variable is much cleaner/easier to type hint this way rather than type hinting the entire kwargs dict (which would be the alternative option). That being said, I'm not sure if you can even type hint the kwargs dict in a way where calling `.get` on it resolves correctly during static analysis.

anttimaki replied:
Assuming there's a downstream (`get_object_or_404`'s kwargs seem to be any-typed).

anttimaki commented on django/thunderstore/frontend/api/experimental/views/community_package_list.py:
```diff
@@ -0,0 +1,253 @@
+    def get(self, request: HttpRequest, **kwargs) -> HttpResponse:
+        communities = Community.objects.prefetch_related("package_categories")
+        community = get_object_or_404(communities, identifier=kwargs.get("community"))
+        package_qs = self.get_queryset(community)
```
I guess more a habit of the view class containing the stuff it needs to produce the result, especially when the methods aren't that reusable.

anttimaki commented on django/thunderstore/frontend/api/experimental/views/community_package_list.py:
```diff
@@ -0,0 +1,253 @@
+        Return cache key for CachedPaginator.
+        """
+        cache_vary = self.kwargs["community"]
+        cache_vary += f".{'deprecated' in request.query_params}"
```
Well I guess if we want to change it, I might as well do the changes now.

anttimaki replied:
Well, it seems the old, non-experimental implementation is no better, since `QueryDict` returns all values as strings (unless I'm missing something), so the only falsy values are a missing key and an empty string. So even if `&deprecated=0` or `&deprecated=false` is sent, it evaluates to `True`.

I can check for empty string (in fact this I've already implemented), but trying to cast other values to correct types seems like a fool's errand. So if we want to have some falsy value, I'd say we should agree on one, and then use it in requests and check for it - and it only - on backend. Thoughts?

anttimaki replied:
```
qp = PackageSearchQueryParameterSerializer(data=request.query_params)
if not qp.is_valid():
    return Response(qp.errors, status=status.HTTP_400_BAD_REQUEST)

params: OrderedDict = qp.validated_data
```

Something like this seems to work so I'll go with it. 👍🏻 for the idea.

anttimaki commented on django/thunderstore/frontend/api/experimental/views/community_package_list.py:
```diff
@@ -0,0 +1,253 @@
+        """
+        if "included_categories" in request.query_params:
+            slugs = request.query_params.getlist("included_categories")
+            queryset = queryset.filter(community_listings__categories__slug__in=slugs)
```
Ah shit, here we go again. Guess I'll switch these to `.exclude()`s. It's NOT NOT gonna be pretty.

anttimaki commented on django/thunderstore/frontend/api/experimental/views/package_details.py:
```diff
@@ -0,0 +1,132 @@
+            self.get_listing_queryset(),
+            community__identifier=kwargs.get("community"),
+            community__is_listed=True,
+            package__name=kwargs.get("package"),
```
Yes, it is, but based on the question I wonder if I'm also missing some pieces of the puzzle here 😅 This endpoint is for frontend URL `/c/communityname/p/packagename/`, so community's ~name~ identifier and package's name is all the information we have available for filtering? How should the browser know which team/namespace to request the package for?

anttimaki replied:
Alright. That sounds a like a bit bigger change (also affecting UI side quite a bit), so I'll leave it last (or implement it in another PR).

MythicManiac replied:
Moved to a separate issue

anttimaki commented on django/thunderstore/frontend/api/experimental/views/package_details.py:
```diff
@@ -0,0 +1,132 @@
+        )
+
+        if not listing.can_be_viewed_by_user(request.user):
+            raise Http404("Package is waiting for approval or has been rejected")
```
Should I only change the status code, or do you think we should change the message to hide the reason for the error as well?

MythicManiac replied:
I think the message should either be derived from the function that checks the permissions, or be omitted. We don't have a way of telling if the message is valid here after all, as the checks are elsewhere. So let's remove the message for now at least.

anttimaki commented on django/thunderstore/frontend/api/experimental/views/package_details.py:
```diff
@@ -0,0 +1,132 @@
+
+            if community is None:
+                msg = f"Dependency {d} of {latest.package} has no PackageListings"
+                raise NotImplementedError(msg)
```
Earlier it was discussed that we should have a community-agnostic view for packages. Should displaying a dependency without links be a temporary solution until that is implemented? Are there reasons we don't want to implemented that view now (e.g. not sure how it should be implemented yet)?

MythicManiac replied:
Yeah that pretty much sums it up 👍

MythicManiac replied:
Although even if we have a non-community associated page, it's possible those too will be unlisted entirely, and we'd still need to be able to show the dependency relation somehow.

anttimaki commented on django/thunderstore/frontend/api/experimental/views/package_details.py:
```diff
@@ -0,0 +1,132 @@
+        Format results to transportation.
+        """
+        latest = listing.package.latest
+        latest.package = listing.package  # To avoid extra DB hits
```
Based on a test done with the query stats decorator included in this PR, two extra queries are done without this hack. One fetches the package itself, and another fetches that package's owner/team.

MythicManiac replied:
Alright, makes sense 👍

anttimaki commented:
Changes based on the feedback:

### Frontpage

* [Unit test](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/316b964b254933735e3d4aba12fc9e7bb72f9a11#diff-11e3d27172a15738549e50934fc7c57f4b2b734a8fcd290038df6db8845faeeaR14) updates rows directly in DB rather than fetching and saving changes in Python

### Community's Package List

* [Pass package_version_kwargs to PackageVersionFactory](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-4fcb659107cfe292c9f74af84f15d50a3afc1e0cca598e8b3b433fe7ef761675R77) instead of populating the fields after the object creation
* Add `is_` prefixes to boolean fields in [serializer definition](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-78ac47ea6d6851af403df02efc91fc817e2a1ffa51e2b1ba944e17466ec972b4R32-R39) and when [serializing results](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-a6757543940dd9330c71cafd21b728b9b2a57129c41e3e0c641a8ba4aa5a8ac9R244-R252)
* Use named `community_identifier` variable in view params instead of reading the value from `kwargs` dict ([view](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-a6757543940dd9330c71cafd21b728b9b2a57129c41e3e0c641a8ba4aa5a8ac9R37), [urls](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-594bf8d7a82e1ec09f3ab6d2fd7550cb68dffc8b79ed7d7ef2c0c9efdb8c69daR11), [tests](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-15ac0f5618a6efabdb9740435542d09b9a716ebf3319d365f9fff9c0ab927c26R453))
* Use [`PackageSearchQueryParameterSerializer`](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-78ac47ea6d6851af403df02efc91fc817e2a1ffa51e2b1ba944e17466ec972b4R65) to [deserialize query parameters](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-a6757543940dd9330c71cafd21b728b9b2a57129c41e3e0c641a8ba4aa5a8ac9R41-R45) for filtering packages (affects also ~all helper methods in the view)
* Changed `.filter`s to `.exlude`s in [here](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-a6757543940dd9330c71cafd21b728b9b2a57129c41e3e0c641a8ba4aa5a8ac9R135-R141) and [here](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-a6757543940dd9330c71cafd21b728b9b2a57129c41e3e0c641a8ba4aa5a8ac9R172-R174)
* Tests:
  * [Use existing invalidate_cache method](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-15ac0f5618a6efabdb9740435542d09b9a716ebf3319d365f9fff9c0ab927c26R8), remove duplicate `bust_cache_by_condition` method
  * Renamed the test from `test_only_community_listings_for_correct_community_are_prefetched` to [`test_only_community_listings_for_correct_community_are_included_in_queryset`](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-15ac0f5618a6efabdb9740435542d09b9a716ebf3319d365f9fff9c0ab927c26R49) to better communicate that this test case only tests the view's `get_queryset` method rather than HTTP response of the view.
  * Added [`test_paginated_results_are_cached`](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-15ac0f5618a6efabdb9740435542d09b9a716ebf3319d365f9fff9c0ab927c26R301)
  * `test_packages_can_be_ordered_by_download_count`: check that [cache busting is actually required in the test](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-15ac0f5618a6efabdb9740435542d09b9a716ebf3319d365f9fff9c0ab927c26R352-R367)
  * `test_pagination`: check that [packages are not reshown on second page](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/3cabd1afc55ff80bbd1a98cd76f666fc8623ec50#diff-15ac0f5618a6efabdb9740435542d09b9a716ebf3319d365f9fff9c0ab927c26R423-R427)

### Package Detail View

* Avoid double iteration via [`x in y` call](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-fb2e291d6e78a485a21aee610c344555db1ffeedc771ab941082d6e40724bcc2R37)
* Use named `community_identifier`/`package_name` variables in view params instead of reading the value from kwargs dict ([view](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-e2fa942fb12d2f38653fcb775852383d85da2c49696d5d6015addd9438fb359aR32), [urls](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-594bf8d7a82e1ec09f3ab6d2fd7550cb68dffc8b79ed7d7ef2c0c9efdb8c69daR17), [tests](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-bf6a3a55bbca65be0733aa924475078e245986c09067dbd0b48b16e611ffa072R222-R223))
* Fetching unapproved package now results in [PermissionDenied](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-e2fa942fb12d2f38653fcb775852383d85da2c49696d5d6015addd9438fb359aR42) error
* Allow unlisted dependencies ([serializer](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-78ac47ea6d6851af403df02efc91fc817e2a1ffa51e2b1ba944e17466ec972b4R61-R62), [serializing](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-e2fa942fb12d2f38653fcb775852383d85da2c49696d5d6015addd9438fb359aR87-R90), [tests](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-bf6a3a55bbca65be0733aa924475078e245986c09067dbd0b48b16e611ffa072R163))
* Tests:
  * Assert same community is returned in [test_dependencies_refer_to_preferred_community](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-bf6a3a55bbca65be0733aa924475078e245986c09067dbd0b48b16e611ffa072R134-R140)
  * Assert different community is returned in [test_dependencies_fallback_to_any_community](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-bf6a3a55bbca65be0733aa924475078e245986c09067dbd0b48b16e611ffa072R153-R159)
* [Parametrized](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-bf6a3a55bbca65be0733aa924475078e245986c09067dbd0b48b16e611ffa072R94-R95) `test_rating_is_annotated_correctly`
* Added [`test_listing_approval_status_is_checked`](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/e919eab69af5daece24761169f1542e6f9a1beeb#diff-bf6a3a55bbca65be0733aa924475078e245986c09067dbd0b48b16e611ffa072R54)

### Other

* Move [debug decorator to a separate package](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/01950443034b9ca26c44a638986b6c7aa4cc69d7#diff-33d606093aceb6297d215f17fde59382289610f254d0e6e52558f462bedb0ab8R8), remove from [code coverage](https://github.com/thunderstore-io/Thunderstore/pull/507/commits/01950443034b9ca26c44a638986b6c7aa4cc69d7#diff-d6639385f8f763e158590830d5b6ae0b4dd2ba5d5e0df7dd15f6609bdc1d4653R13)

### Not done

* Anything related to unresolved comments
* Related changes in UI repo

MythicManiac reviewed (commented):
Got a few more comments but overall looks good

MythicManiac commented on django/thunderstore/frontend/api/experimental/serializers/views.py:
```diff
@@ -0,0 +1,131 @@
+    download_count = serializers.IntegerField(min_value=0)
+    download_url = serializers.CharField()
+    install_url = serializers.CharField()
+    version_number = serializers.CharField(max_length=16)
```
This is used elsewhere at least:
```python
    name = models.CharField(
        max_length=Package._meta.get_field("name").max_length,
    )
```

xiaoxiao921 commented:
is there any reason the usage of ModelSerializer is not more widespread, its used nicely with the PackageCategorySerializer but not on cases where it would've saved a lot of typing / model duplication like the PackageDetailViewContentSerializer or PackageVersionSerializer for example.

MythicManiac reviewed (approved):
LGTM 👍 nice work

xiaoxiao921 commented:
I said this because most of the fields were 1:1 matching, you can still override specific fields as needed

MythicManiac commented:
> I said this because most of the fields were 1:1 matching, you can still override specific fields as needed

Yeah, the trouble is how changes to models can lead to unintentional changes in the API. Of course tests can be written checking for this, but separately declaring the API surface makes them less tightly coupled.

I don't think there's really a right or wrong way to go about it here, just my experience based gut feeling saying this is approach leads to less issues long-term even if it's slightly more code. There are scenarios where code duplication makes a lot of sense when thinking about the implications, and this I feel like is one of those scenarios.

So in short, design choice based on my experience, some intuitive reasoning above.
